sdk/test-utils/perf/src/batchPerfTest.ts

```diff
@@ -0,0 +1,199 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { AbortSignalLike, AbortController } from "@azure/abort-controller";
+import {
+  TestProxyHttpClient,
+  TestProxyHttpClientV1,
+  testProxyHttpPolicy
+} from "./testProxyHttpClient";
+import { HttpClient } from "@azure/core-http";
+import { Pipeline } from "@azure/core-rest-pipeline";
+import { PerfTestBase } from "./perfTestBase";
+import { PerfParallel } from "./parallel";
+
+/**
+ * Enables writing perf tests where the number of operations are dynamic for the method/call being tested.
+ */
+export abstract class BatchPerfTest<TOptions = Record<string, unknown>> extends PerfTestBase<
+  TOptions
+> {
+  private readonly testProxy!: string;
+  public testProxyHttpClient!: TestProxyHttpClient;
+  public testProxyHttpClientV1!: TestProxyHttpClientV1;
+
+  public constructor() {
+    super();
+    const testProxies = this.parsedOptions["test-proxies"].value;
+    if (testProxies) {
+      const testProxiesArray = testProxies.split(";");
+      this.testProxy = testProxiesArray[this.parallelIndex % testProxiesArray.length];
+    }
+  }
+
+  public abstract runBatch(abortSignal?: AbortSignalLike): Promise<number>;
+
+  /**
+   * configureClientOptionsCoreV1
+   *
+   * For core-v1 - libraries depending on core-http
+   * Apply this method on the client options to get the proxy tool support
+   *
+   * Note: httpClient must be part of the options bag, it is required for the perf framework to update the underlying client properly
+   */
+  public configureClientOptionsCoreV1<T>(options: T & { httpClient?: HttpClient }): T {
+    if (this.testProxy) {
+      this.testProxyHttpClientV1 = new TestProxyHttpClientV1(
+        this.testProxy,
+        this.parsedOptions["insecure"].value
+      );
+      options.httpClient = this.testProxyHttpClientV1;
+    }
+    return options;
+  }
+
+  /**
+   * configureClient
+   *
+   * For core-v2 - libraries depending on core-rest-pipeline
+   * Apply this method on the client to get the proxy tool support.
+   *
+   * Note: Client must expose the pipeline property which is required for the perf framework to add its policies correctly
+   */
+  public configureClient<T>(client: T & { pipeline: Pipeline }): T {
+    if (this.testProxy) {
+      this.testProxyHttpClient = new TestProxyHttpClient(
+        this.testProxy,
+        this.parsedOptions["insecure"].value ?? false
+      );
+      client.pipeline.addPolicy(
+        testProxyHttpPolicy(
+          this.testProxyHttpClient,
+          this.testProxy.startsWith("https"),
+          this.parsedOptions["insecure"].value ?? false
+        )
+      );
+    }
+    return client;
+  }
+
+  /**
+   * Runs the test in scope repeatedly, without waiting for any promises to finish,
```

Comment:
> waiting for any promises to finish

I am confused by this comment because the only promise in scope, `this.runBatch(abortController.signal)`, is being awaited.